Return None correctly from `Tensor.names`

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28659 Return None correctly from `Tensor.names`**

Previously, we would return None from `Tensor.names` without bumping the
refcount. This is a bug: PyTuple_SET_ITEM steals a reference to the object.
When the tuple is deallocated, it'll decrement the refcount on Py_None.
The end result is that if Tensor.names is called a lot, Py_None's
refcount hits 0 and causes Python to throw an error.

To avoid this, we "create" a new reference to Py_None by increasing
the refcount and let PyTuple_SET_ITEM steal that reference.

See the following for Python documentation on this:
- https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SetItem
- https://stackoverflow.com/questions/16400600/how-to-return-a-tuple-containing-a-none-value-from-the-c-api

Fixes https://github.com/pytorch/pytorch/issues/28646

Test Plan:
- New test.

Differential Revision: [D18140593](https://our.internmc.facebook.com/intern/diff/D18140593)

zou3519 committed:
Return None correctly from `Tensor.names`

Previously, we would return None from `Tensor.names` without bumping the
refcount. This is a bug; the Python API requires the developer to
increment the refcount on new references to None. This is because None
is a singleton object and does not automatically have its reference
count bumped when one uses Py_None (which is a pointer to the actual
None singleton object).

See the following for Python documentation on this:
- https://docs.python.org/3/c-api/none.html#c.Py_RETURN_NONE
- https://docs.python.org/3/extending/extending.html#back-to-the-example

Fixes https://github.com/pytorch/pytorch/issues/28646

Test Plan:
- New test.

[ghstack-poisoned]

zou3519 committed:
Update on "Return None correctly from `Tensor.names`"


Previously, we would return None from `Tensor.names` without bumping the
refcount. This is a bug; the Python API requires the developer to
increment the refcount on new references to None. This is because None
is a singleton object and does not automatically have its reference
count bumped when one uses Py_None (which is a pointer to the actual
None singleton object).

See the following for Python documentation on this:
- https://docs.python.org/3/c-api/none.html#c.Py_RETURN_NONE
- https://docs.python.org/3/extending/extending.html#back-to-the-example

Fixes https://github.com/pytorch/pytorch/issues/28646

Test Plan:
- New test.

[ghstack-poisoned]

gchanan commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -335,8 +335,17 @@ PyObject *THPVariable_get_names(THPVariable *self, void *unused)
+      // See https://docs.python.org/3/extending/extending.html#back-to-the-example
+      // and https://docs.python.org/3/c-api/none.html#c.Py_RETURN_NONE
+      // Note that we cannot use Py_RETURN_NONE directly because we return a list
+      // of names, of which some can be None.
```
> we need to increment the refcount because this is a *new* reference.

I would phrase the actual issue as `PyTuple_SET_ITEM` steals the reference.

zou3519 replied:
Thanks for catching that, I didn't realize that PyTuple_SET_ITEM is what takes the reference

zou3519 committed:
Update on "Return None correctly from `Tensor.names`"


Previously, we would return None from `Tensor.names` without bumping the
refcount. This is a bug; the Python API requires the developer to
increment the refcount on new references to None. This is because None
is a singleton object and does not automatically have its reference
count bumped when one uses Py_None (which is a pointer to the actual
None singleton object).

See the following for Python documentation on this:
- https://docs.python.org/3/c-api/none.html#c.Py_RETURN_NONE
- https://docs.python.org/3/extending/extending.html#back-to-the-example

Fixes https://github.com/pytorch/pytorch/issues/28646

Test Plan:
- New test.

Differential Revision: [D18140593](https://our.internmc.facebook.com/intern/diff/D18140593)

[ghstack-poisoned]

zou3519 committed:
Update on "Return None correctly from `Tensor.names`"


Previously, we would return None from `Tensor.names` without bumping the
refcount. This is a bug: PyTuple_SET_ITEM steals a reference to the object.
When the tuple is deallocated, it'll decrement the refcount on Py_None.
The end result is that if Tensor.names is called a lot, Py_None's
refcount hits 0 and causes Python to throw an error.

To avoid this, we "create" a new reference to Py_None by increasing
the refcount and let PyTuple_SET_ITEM steal that reference.

See the following for Python documentation on this:
- https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SetItem
- https://stackoverflow.com/questions/16400600/how-to-return-a-tuple-containing-a-none-value-from-the-c-api

Fixes https://github.com/pytorch/pytorch/issues/28646

Test Plan:
- New test.

Differential Revision: [D18140593](https://our.internmc.facebook.com/intern/diff/D18140593)

[ghstack-poisoned]
